docs/source/schemas/dynamic_columns.rst

```diff
@@ -0,0 +1,263 @@
+.. code:: java
+
+    schema.addTableDefinition("todo", new TableDefinition() {{
+                rowName();
+                    rowComponent("person", ValueType.STRING);
+                dynamicColumns();
+                    columnComponent("taskSize", ValueType.VAR_LONG);
+                    columnComponent("monetaryCost", ValueType.VAR_LONG);
+                    value(ValueType.STRING);
+            }});
+
+Note that dynamic column components must be primitive ValueTypes which support partitioning and ordering.
+Also, key-value pairs in an individual row will be returned in *lexicographically* sorted order of the key.
+
+We call the person the *row key*, the (taskSize, monetaryCost) pair the *dynamic column key*, and the string (which is
+intended to be a description of the todo) the *value*.
+
+Query Capabilities
+------------------
+
+Dynamic Columns are useful for avoiding KVS-level range scans, especially in key-value services where range scans
+are expensive (like Cassandra - its caching optimisations are rendered ineffectual for range scans). For example,
+for the schema defined above:
+
+1. "Find John's smallest and cheapest todo" can be answered very efficiently.
+   Elements are stored in sorted order, so the very first key-value pair retrieved from the database is actually the
+   correct answer to this query.
+2. "Find all of Tom's todos with size up to M, limited to N results" is easy, as this can be
+   processed with a natural stopping point, as in query 1, and we are iterating in the right order for this query.
+   This can be achieved using the getRowsColumnRange API.
+3. "Find all of Tom's todos with size of 10 to 15, limit to N results" can be performed via a range
+   scan on the dynamic column key. This is also readily supported, and does not create a KVS-level range scan.
+4. "Find all of Tom's largest todos" is somewhat more difficult, as reverse range scans aren't supported by the
+   dynamic columns API. If we were more interested in largest todos as opposed to smallest, we could set the
+   ``ValueByteOrder`` of the ``taskSize`` column component to decreasing.
+5. "Find all of Tom's todos which cost less than 5" will be less efficient. We need to retrieve the entire row to do
+   this, because the map is sorted by task size and then monetary cost. If queries on cost are more common than queries
+   on size, flipping the order of the column components could help.
+6. "Find all of Tom's todos which have size between 3 and 7, and cost between 5 and 10". This is achievable, though in
+   addition to the values we want to see, we also need to scan through all values with size 4-6, and those with size
+   3 and cost >10, or 7 and cost <5.
+7. "Find all of Nathan's todos that begin with the letter N". Prefix / range scans are not supported on values, so this
+   requires us to scan through the entire row.
+8. "Find all of John's and Jeremy's smallest todos" is easy for a similar reason to 1. We provide an API that allows
```

Comment:
Actually I need to clarify this. This was _originally_ the same as doing 1 twice, but no longer is because I changed the schema to have a two element dynamic column key (and this is only on the first element, while query 1 was on both elements of the key). I could mention that it is the same as doing the query "find a person's smallest todos" twice.